File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.cs
public class Solution {
    private bool Possible(int[] arr, int day, int m, int k) {
        int n = arr.Length;
        int cnt = 0;
        int noOfBouquets = 0;

        // Count the number of bouquets
        for (int i = 0; i < n; i++) {
            if (arr[i] <= day) {
                cnt++;
            } else {
                noOfBouquets += (cnt / k);
                cnt = 0;
            }
        }
        noOfBouquets += (cnt / k);
        return noOfBouquets >= m;
    }

    private int RoseGarden(int[] arr, int k, int m) {
        long val = (long)m * k;
        int n = arr.Length;
        if (val > n) return -1;

        // Find maximum and minimum
        int mini = int.MaxValue, maxi = int.MinValue;
        for (int i = 0; i < n; i++) {
            mini = Math.Min(mini, arr[i]);
            maxi = Math.Max(maxi, arr[i]);
        }

        // Apply binary search
        int low = mini, high = maxi;
        while (low <= high) {
            int mid = (low + high) / 2;
            if (Possible(arr, mid, m, k)) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return low;
    }
    public int MinDays(int[] bloomDay, int m, int k) {
        return RoseGarden(bloomDay, k, m);
    }
}

File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.cpp
class Solution {
public:
    bool possible(vector<int> &arr, int day, int m, int k) {
        int n = arr.size(); //size of the array
        int cnt = 0;
        int noOfB = 0;
        // count the number of bouquets:
        for (int i = 0; i < n; i++) {
            if (arr[i] <= day) {
                cnt++;
            }
            else {
                noOfB += (cnt / k);
                cnt = 0;
            }
        }
        noOfB += (cnt / k);
        return noOfB >= m;
    }
    int roseGarden(vector<int> arr, int k, int m) {
        long long val = m * 1ll * k * 1ll;
        int n = arr.size(); //size of the array
        if (val > n) return -1; //impossible case.
        //find maximum and minimum:
        int mini = INT_MAX, maxi = INT_MIN;
        for (int i = 0; i < n; i++) {
            mini = min(mini, arr[i]);
            maxi = max(maxi, arr[i]);
        }

        //apply binary search:
        int low = mini, high = maxi;
        while (low <= high) {
            int mid = (low + high) / 2;
            if (possible(arr, mid, m, k)) {
                high = mid - 1;
            }
            else low = mid + 1;
        }
        return low;
    }
    int minDays(vector<int>& bloomDay, int m, int k) {
        // int cnt = 0;
        // int n = bloomDay.size();
        // int minDay = 0;
        // for(int i=0; i<n; i++)
        // {
        //     if(bloomDay[i]>minDay)
        //     {
        //         minDay = bloomDay[i];
        //     }
        // }
        // bool flag = false;
        // int ans = INT_MAX;
        // while(minDay)
        // {
        //     int adjacent = 0;
        //     cnt = 0;
        //     for(int i=0; i<n; i++)
        //     {
        //         if(bloomDay[i]<=minDay)
        //         {
        //             adjacent++;
        //             if(adjacent==k)
        //             {
        //                 cnt++;
        //                 adjacent = 0;
        //             } 
        //         }
        //         else
        //         {
        //             adjacent = 0;
        //         }
        //         if(cnt==m)
        //         {
        //             flag = true;
        //             ans = min(ans,minDay);
        //             break;
        //         }
        //     }
        //     minDay--;
        // }
        // return flag==true?ans:-1;
        return roseGarden(bloomDay, k, m);
    }
};